File: python-app/models/image_slide/self.py
from models.slide.self import Slide
from models.slide.slide_category import SlideCategory
from models.image_slide.image_slide_kind import ImageSlideKind
from models.image_slide.image_slide_parameter import ImageSlideParameter

from render.drawable_area import DrawableArea

from abc import abstractmethod

class ImageSlide(Slide):
    def __init__(self, title, identifier, creation_date, last_edit, preview, kind: ImageSlideKind):
        super().__init__(title, identifier, creation_date, last_edit, preview, SlideCategory.IMAGE_SLIDE)
        self.kind = kind
    
    def to_dict(self):
        return {
            **super().to_dict(),
            'kind': self.kind.value,
            'parameters': {
                key: value.to_dict() for (key, value) in self.parameters_dict.items()
            }
        }

    @property
    @abstractmethod
    def parameters_dict(self) -> dict[str, ImageSlideParameter]:
        pass

    @abstractmethod
    def render(self, slide: Slide, drawable_area: DrawableArea):
        pass

File: lib/models/image_slide/self.dart
import 'package:ut_report_generator/models/image_slide/image_slide_kind.dart';
import 'package:ut_report_generator/models/image_slide/image_slide_parameter.dart';
import 'package:ut_report_generator/models/slide/self.dart';
import 'package:ut_report_generator/models/slide_category.dart';

class ImageSlide extends Slide {
  ImageSlideKind kind;
  Map<String, ImageSlideParameter> parameters;

  ImageSlide({
    required super.identifier,
    required super.title,
    required super.creationDate,
    required super.lastEdit,
    required super.preview,
    required this.kind,
    required this.parameters,
  }) : super(category: SlideCategory.imageSlide);

  @override
  Map<String, dynamic> toMap() {
    return {
      ...super.toMap(),
      'kind': kind.name,
      'parameters': parameters.map(
        (key, value) => MapEntry(key, value.toMap()),
      ),
    };
  }

  factory ImageSlide.fromJson(Map<String, dynamic> json) {
    return ImageSlide(
      identifier: json["identifier"],
      title: json["title"],
      creationDate: DateTime.parse(json["creation_date"]),
      lastEdit: DateTime.parse(json["last_edit"]),
      preview: json["preview"],
      kind: ImageSlideKind.values.byName(json["kind"]),
      parameters: (json["parameters"] as Map<String, dynamic>).map(
        (key, value) => MapEntry(key, ImageSlideParameter.fromJson(value)),
      ),
    );
  }
}
